File: Design-Patterns-Using-Java/Decorator/DecoratorPatternCoffee.java
// Coffee (Component)
interface Coffee {
    int getCost();
    String getIngredients();
}

// SimpleCoffee (ConcreteComponent)
class SimpleCoffee implements Coffee {
    @Override
    public int getCost() {
        return 10;
    }

    @Override
    public String getIngredients() {
        return "Coffee";
    }
}

// CoffeeDecorator (Decorator)
abstract class CoffeeDecorator implements Coffee {
    protected Coffee component;

    public CoffeeDecorator(Coffee component) {
        this.component = component;
    }

    @Override
    public abstract int getCost();

    @Override
    public abstract String getIngredients();
}

// MilkDecorator (ConcreteDecorator)
class MilkDecorator extends CoffeeDecorator {
    public MilkDecorator(Coffee component) {
        super(component);
    }

    @Override
    public int getCost() {
        return component.getCost() + 4;
    }

    @Override
    public String getIngredients() {
        return component.getIngredients() + ", Milk";
    }
}

// EspressoDecorator (ConcreteDecorator)
class EspressoDecorator extends CoffeeDecorator {
    public EspressoDecorator(Coffee component) {
        super(component);
    }

    @Override
    public int getCost() {
        return component.getCost() + 5;
    }

    @Override
    public String getIngredients() {
        return component.getIngredients() + ", Espresso";
    }
}

// Client code
public class DecoratorPatternCoffee {
    public static void main(String[] args) {
        Coffee component = new SimpleCoffee();
        Coffee decorator1 = new MilkDecorator(component);
        Coffee decorator2 = new EspressoDecorator(decorator1);

        System.out.println("Coffee cost is :: " + decorator2.getCost());
        System.out.println("Coffee ingredients are :: " + decorator2.getIngredients());

        Coffee latte = new MilkDecorator(new MilkDecorator(new SimpleCoffee()));
        System.out.println("Coffee cost is :: " + latte.getCost());
        System.out.println("Coffee ingredients are :: " + latte.getIngredients());
    }
}

/*
Coffee cost is :: 19
Coffee ingredients are :: Coffee, Milk, Espresso
Coffee cost is :: 18
Coffee ingredients are :: Coffee, Milk, Milk
*/

File: Design-Patterns-Using-CSharp/Decorator/DecoratorPatternCoffee.cs
using System;

// Coffee (Component)
interface ICoffee
{
    int GetCost();
    string GetIngredients();
}

// SimpleCoffee (ConcreteComponent)
class SimpleCoffee : ICoffee
{
    public int GetCost()
    {
        return 10;
    }

    public string GetIngredients()
    {
        return "Coffee";
    }
}

// CoffeeDecorator (Decorator)
abstract class CoffeeDecorator : ICoffee
{
    protected ICoffee Component;

    public CoffeeDecorator(ICoffee component)
    {
        Component = component ?? throw new ArgumentNullException(nameof(component));
    }

    public abstract int GetCost();

    public abstract string GetIngredients();
}

// MilkDecorator (ConcreteDecorator)
class MilkDecorator : CoffeeDecorator
{
    public MilkDecorator(ICoffee component) : base(component)
    {
    }

    public override int GetCost()
    {
        return Component.GetCost() + 4;
    }

    public override string GetIngredients()
    {
        return Component.GetIngredients() + ", Milk";
    }
}

// EspressoDecorator (ConcreteDecorator)
class EspressoDecorator : CoffeeDecorator
{
    public EspressoDecorator(ICoffee component) : base(component)
    {
    }

    public override int GetCost()
    {
        return Component.GetCost() + 5;
    }

    public override string GetIngredients()
    {
        return Component.GetIngredients() + ", Espresso";
    }
}

// Client code
class DecoratorPatternCoffee
{
    static void Main(string[] args)
    {
        ICoffee component = new SimpleCoffee();
        ICoffee decorator1 = new MilkDecorator(component);
        ICoffee decorator2 = new EspressoDecorator(decorator1);

        Console.WriteLine("Coffee cost is :: " + decorator2.GetCost());
        Console.WriteLine("Coffee ingredients are :: " + decorator2.GetIngredients());

        ICoffee latte = new MilkDecorator(new MilkDecorator(new SimpleCoffee()));
        Console.WriteLine("Coffee cost is :: " + latte.GetCost());
        Console.WriteLine("Coffee ingredients are :: " + latte.GetIngredients());
    }
}

/*
Coffee cost is :: 19
Coffee ingredients are :: Coffee, Milk, Espresso
Coffee cost is :: 18
Coffee ingredients are :: Coffee, Milk, Milk
*/